From dd389b58c7e5b18ab396fdf257c331421a801046 Mon Sep 17 00:00:00 2001
From: Hanna Hawa <hannah@marvell.com>
Date: Sun, 3 Aug 2014 22:52:39 +0300
Subject: [PATCH 0048/1240] a38x: mpp: update the MPPs list for A38x SoC

	- update the static MPPs for RD & DB board
	- add full MPPs description for A38x
	- add initial MPP busses

Signed-off-by: Hanna Hawa <hannah@marvell.com>
Change-Id: I8988e78a12ed13361fc52580d0a4da26821d561d
Reviewed-on: http://vgitil04.il.marvell.com:8080/9802
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Yehuda Yitschak <yehuday@marvell.com>
---
 arch/arm/cpu/armv7/armada38x/tables.c   | 79 ++++++++++++++++++++++++++++-----
 arch/arm/include/asm/arch-mvebu/mpp.h   |  7 ++-
 board/mvebu/armada38x/db-88f6820-info.c |  4 +-
 board/mvebu/armada38x/rd-88f6820-info.c |  4 +-
 board/mvebu/common/init.c               |  2 +-
 5 files changed, 80 insertions(+), 16 deletions(-)

diff --git a/arch/arm/cpu/armv7/armada38x/tables.c b/arch/arm/cpu/armv7/armada38x/tables.c
index 7b376a3..fa5c445 100644
--- a/arch/arm/cpu/armv7/armada38x/tables.c
+++ b/arch/arm/cpu/armv7/armada38x/tables.c
@@ -64,20 +64,79 @@ u16 armada6285_unit_disable[MAX_UNIT_ID] = {
 /* MPP description table. Informative only */
 char *a38x_mpp_desc[MAX_MPP_ID + 1][MAX_MPP_OPTS] = {
 /*	0x0        0x1         0x2              0x3             0x4             0x5             0x6 */
-/* 0 */	{"GPIO_0", "UA_0_RXD", "NA",		"NA",		"NA",           "NA",		"NA"},
-/* 1 */	{"GPIO_1", "UA_0_TXD", "NA",		"NA",		"NA",		"NA",		"NA"},
-/* 2 */	{"GPIO_2", "I2C_0_SCK", "NA",		"NA",		"NA",		"NA",		"NA"},
-/* 3 */	{"GPIO_3", "I2C_0_SDA", "NA",		"NA",		"NA",		"NA",		"NA"},
-/* 4 */	{"GPIO_4", "GE_MDC",    "UA_1_RXD",	"UA_0_RTS",	"NA",		"NA",		"NA"},
-/* 5 */	{"GPIO_5", "GE_MDIO",   "UA_1_TXD",	"UA_0_CTS",	"NA",		"NA",		"NA"},
+/* 0 */ { "GPIO[0]", "UA0_RXD", "NA", "NA", "NA", "NA", "NA"},
+/* 1 */ { "GPIO[1]", "UA0_TXD", "NA", "NA", "NA", "NA", "NA"},
+/* 2 */ { "GPIO[2]", "I2C0_SCK", "NA", "NA", "NA", "NA", "NA"},
+/* 3 */ { "GPIO[3]", "I2C0_SDA", "NA", "NA", "NA", "NA", "NA"},
+/* 4 */ { "GPIO[4]", "GE_MDC", "UA1_TXD", "UA0_RTS", "NA", "NA", "NA"},
+/* 5 */ { "GPIO[5]", "GE_MDIO", "UA1_RXD", "UA0_CTS", "NA", "NA", "NA"},
+/* 6 */ { "GPIO[6]", "GE0_TXCLKOUT", "GE0_CRS", "NA", "NA", "DEV_CSn[3]", "NA"},
+/* 7 */ { "GPIO[7]", "GE0_TXD[0]", "NA", "NA", "NA", "DEV_AD[9]", "NA"},
+/* 8 */ { "GPIO[8]", "GE0_TXD[1]", "NA", "NA", "NA", "DEV_AD[10]", "NA"},
+/* 9 */ { "GPIO[9]", "GE0_TXD[2]", "NA", "NA", "NA", "DEV_AD[11]", "NA"},
+/* 10 */ { "GPIO[10]", "GE0_TXD[3]", "NA", "NA", "NA", "DEV_AD[12]", "NA"},
+/* 11 */ { "GPIO[11]", "GE0_TXCTL", "NA", "NA", "NA", "DEV_AD[13]", "NA"},
+/* 12 */ { "GPIO[12]", "GE0_RXD[0]", "PCIe0_RSTOUTn", "PCIe1_RSTOUTn", "SPI0_CSn[1]", "DEV_AD[14]", "PCIe3_CLKREQ"},
+/* 13 */ { "GPIO[13]", "GE0_RXD[1]", "PCIe0_CLKREQ", "PCIe1_CLKREQ", "SPI0_CSn[2]", "DEV_AD[15]", "PCIe2_CLKREQ"},
+/* 14 */ { "GPIO[14]", "GE0_RXD[2]", "PTP_CLK", "M_VTT_CTRL", "SPI0_CSn[3]", "DEV_WEn[1]", "PCIe3_CLKREQ"},
+/* 15 */ { "GPIO[15]", "GE0_RXD[3]", "GE_MDC Slave", "PCIe0_RSTOUTn", "SPI0_MOSI", "PCIe1_RSTOUTn", "PCIe2_RSTOUTn"},
+/* 16 */ { "GPIO[16]", "GE0_RXCTL", "GE_MDIO Slave", "M_DECC_ERR", "SPI0_MISO", "PCIe0_CLKREQ", "PCIe1_CLKREQ"},
+/* 17 */ { "GPIO[17]", "GE0_RXCLK", "PTP_CLK", "UA1_RXD", "SPI0_SCK", "SATA1_PRESENT_ACTIVEn", "SATA0_PRESENT_ACTIVEn"},
+/* 18 */ { "GPIO[18]", "GE0_RXERR", "PTP_TRIG_GEN", "UA1_TXD", "SPI0_CSn[0]", "PCIe1_RSTOUTn", "PCIe2_RSTOUTn"},
+/* 19 */ { "GPIO[19]", "GE0_COL", "PTP_EVENT_REQ", "PCIe0_CLKREQ", "SATA1_PRESENT_ACTIVEn", "UA0_CTS", "UA1_RXD"},
+/* 20 */ { "GPIO[20]", "GE0_TXCLK", "PTP_CLK", "PCIe1_RSTOUTn", "SATA0_PRESENT_ACTIVEn", "UA0_RTS", "UA1_TXD"},
+/* 21 */ { "GPIO[21]", "SPI0_CSn[1]", "GE1_RXD[0]", "SATA0_PRESENT_ACTIVEn", "SD0_CMD", "DEV_BOOTCSn", "SATA1_PRESENT_ACTIVEn"},
+/* 22 */ { "GPIO[22]", "SPI0_MOSI", "NA", "NA", "NA", "DEV_AD[0]", "NA"},
+/* 23 */ { "GPIO[23]", "SPI0_SCK", "NA", "NA", "NA", "DEV_AD[2]", "NA"},
+/* 24 */ { "GPIO[24]", "SPI0_MISO", "UA0_CTS", "UA1_RXD", "SD0_D[4]", "DEV_READYn", "NA"},
+/* 25 */ { "GPIO[25]", "SPI0_CSn[0]", "UA0_RTS", "UA1_TXD", "SD0_D[5]", "DEV_CSn[0]", "NA"},
+/* 26 */ { "GPIO[26]", "SPI0_CSn[2]", "NA", "I2C1_SCK", "SD0_D[6]", "DEV_CSn[1]", "NA"},
+/* 27 */ { "GPIO[27]", "SPI0_CSn[3]", "GE1_TXCLKOUT", "I2C1_SDA", "SD0_D[7]", "DEV_CSn[2]", "NA"},
+/* 28 */ { "GPIO[28]", "NA", "GE1_TXD[0]", "NA", "SD0_CLK", "DEV_AD[5]", "NA"},
+/* 29 */ { "GPIO[29]", "NA", "GE1_TXD[1]", "NA", "NA", "DEV_ALE[0]", "NA"},
+/* 30 */ { "GPIO[30]", "NA", "GE1_TXD[2]", "NA", "NA", "DEV_OEn", "NA"},
+/* 31 */ { "GPIO[31]", "NA", "GE1_TXD[3]", "NA", "NA", "DEV_ALE[1]", "NA"},
+/* 32 */ { "GPIO[32]", "NA", "GE1_TXCTL", "NA", "NA", "DEV_WEn[0]", "NA"},
+/* 33 */ { "GPIO[33]", "M_DECC_ERR", "NA", "NA", "NA", "DEV_AD[3]", "NA"},
+/* 34 */ { "GPIO[34]", "NA", "NA", "NA", "NA", "DEV_AD[1]", "NA"},
+/* 35 */ { "GPIO[35]", "REF_CLK_OUT[1]", "NA", "NA", "NA", "DEV_A[1]", "NA"},
+/* 36 */ { "GPIO[36]", "PTP_TRIG_GEN", "NA", "NA", "NA", "DEV_A[0]", "NA"},
+/* 37 */ { "GPIO[37]", "PTP_CLK", "GE1_RXCLK", "NA", "SD0_D[3]", "DEV_AD[8]", "NA"},
+/* 38 */ { "GPIO[38]", "PTP_EVENT_REQ", "GE1_RXD[1]", "REF_CLK_OUT[0]", "SD0_D[0]", "DEV_AD[4]", "NA"},
+/* 39 */ { "GPIO[39]", "I2C1_SCK", "GE1_RXD[2]", "UA0_CTS", "SD0_D[1]", "DEV_A[2]", "NA"},
+/* 40 */ { "GPIO[40]", "I2C1_SDA", "GE1_RXD[3]", "UA0_RTS", "SD0_D[2]", "DEV_AD[6]", "NA"},
+/* 41 */ { "GPIO[41]", "UA1_RXD", "GE1_RXCTL", "UA0_CTS", "SPI1_CSn[3]", "DEV_BURSTnLASTn", "ND_RBn[0]"},
+/* 42 */ { "GPIO[42]", "UA1_TXD", "NA", "UA0_RTS", "NA", "DEV_AD[7]", "NA"},
+/* 43 */ { "GPIO[43]", "PCIe0_CLKREQ", "M_VTT_CTRL", "M_DECC_ERR", "PCIe0_RSTOUTn", "DEV_CLKOUT", "ND_RBn[1]"},
+/* 44 */ { "GPIO[44]", "SATA0_PRESENT_ACTIVEn", "SATA1_PRESENT_ACTIVEn", "NA", "NA", "PCIe0_RSTOUTn", "PCIe1_RSTOUTn"},
+/* 45 */ { "GPIO[45]", "REF_CLK_OUT[0]", "PCIe0_RSTOUTn", "PCIe1_RSTOUTn", "PCIe2_RSTOUTn", "PCIe3_RSTOUTn", "UA1_RXD"},
+/* 46 */ { "GPIO[46]", "REF_CLK_OUT[1]", "PCIe0_RSTOUTn", "PCIe1_RSTOUTn", "PCIe2_RSTOUTn", "PCIe3_RSTOUTn", "UA1_TXD"},
+/* 47 */ { "GPIO[47]", "SATA0_PRESENT_ACTIVEn", "SATA1_PRESENT_ACTIVEn", "NA", "SPI1_CSn[2]", "NA", "PCIe2_RSTOUTn"},
+/* 48 */ { "GPIO[48]", "SATA0_PRESENT_ACTIVEn", "M_VTT_CTRL", "TDM2C_PCLK", "AU_I2SMCLKSPDIFRMCLK", "SD0_D[4]", "PCIe0_CLKREQ"},
+/* 49 */ { "GPIO[49]", "NA", "NA", "TDM2C_FSYNC", "AU_I2SLRCLK", "SD0_D[5]", "PCIe1_CLKREQ"},
+/* 50 */ { "GPIO[50]", "PCIe0_RSTOUTn", "PCIe1_RSTOUTn", "TDM2C_DRX", "AU_EXTCLK", "SD0_CMD", "PCIe2_RSTOUTn"},
+/* 51 */ { "GPIO[51]", "NA", "NA", "TDM2C_DTX", "AU_I2SDOAU_SPDIFO", "M_DECC_ERR", "PTP_TRIG_GEN"},
+/* 52 */ { "GPIO[52]", "PCIe0_RSTOUTn", "PCIe1_RSTOUTn", "TDM2C_INTn", "AU_I2SDI", "SD0_D[6]", "PTP_CLK"},
+/* 53 */ { "GPIO[53]", "SATA1_PRESENT_ACTIVEn", "SATA0_PRESENT_ACTIVEn", "TDM2C_RSTn", "AU_I2SBCLK", "SD0_D[7]", "PTP_EVENT_REQ"},
+/* 54 */ { "GPIO[54]", "SATA0_PRESENT_ACTIVEn", "SATA1_PRESENT_ACTIVEn", "PCIe0_RSTOUTn", "PCIe1_RSTOUTn", "SD0_D[3]", "PCIe2_RSTOUTn"},
+/* 55 */ { "GPIO[55]", "UA1_CTS", "GE_MDIO", "PCIe1_CLKREQ", "SPI1_CSn[1]", "SD0_D[0]", "UA1_RXD"},
+/* 56 */ { "GPIO[56]", "UA1_RTS", "GE_MDC", "M_DECC_ERR", "SPI1_MOSI", "NA", "UA1_TXD"},
+/* 57 */ { "GPIO[57]", "NA", "NA", "NA", "SPI1_SCK", "SD0_CLK", "UA1_TXD"},
+/* 58 */ { "GPIO[58]", "PCIe1_CLKREQ", "I2C1_SCK", "PCIe2_CLKREQ", "SPI1_MISO", "SD0_D[1]", "UA1_RXD"},
+/* 59 */ { "GPIO[59]", "PCIe0_RSTOUTn", "I2C1_SDA", "PCIe1_RSTOUTn", "SPI1_CSn[0]", "SD0_D[2]", "PCIe2_RSTOUTn"},
 };
 
 /* Describe the available MPP buses */
 struct mpp_bus a38x_mpp_buses[MAX_MPP_BUS] = {
-	[UART_0_MPP_BUS] = {"UART 0", 2, 1, {{{0, 0x1}, {1, 0x1} } } },
-	[UART_1_MPP_BUS] = {"UART 1", 2, 1, {{{4, 0x2}, {5, 0x2} } } },
-	[SPI_0_MPP_BUS]  = {"SPI 0", 4, 1, {{{0, 0x1}, {1, 0x1} } } },
-	[SPI_1_MPP_BUS]  = {"SPI 1", 4, 1, {{{0, 0x1}, {1, 0x1} } } },
+	[UART_0_MPP_BUS] = {"UART 0", 2, 1, {{{0, 0x1 }, {0, 0x1} } } },
+	[UART_1_MPP_BUS] = {"UART 1", 2, 2, {{{4, 0x2}, {5, 0x2} }, {{19, 0x6}, {20, 0x6} } } },
+	[I2C_0_MPP_BUS] = {"I2C 0", 2, 1, {{{2, 0x1}, {3, 0x1} } } },
+	[GE_0_MPP_BUS] = {"GE 0", 14, 1, {{{4, 0x1}, {5, 0x1}, {6, 0x1}, {7, 0x1}, {8, 0x1}, {9, 0x1}, {10, 0x1}, {11, 0x1}, \
+										{12, 0x1}, {13, 0x1}, {14, 0x1}, {15, 0x1}, {16, 0x1}, {17, 0x1} } } },
+	[GE_1_MPP_BUS] = {"GE 1", 12, 1, {{{21, 0x2}, {27, 0x2}, {28, 0x2}, {29, 0x2}, {30, 0x2}, {31, 0x2}, {32, 0x2}, {37, 0x2}, {38, 0x2}, {39, 0x2}, {40, 0x2}, {41, 0x2} } } },
+	[SPI_0_MPP_BUS]  = {"SPI 0", 4, 1, {{{22, 0x1}, {23, 0x1}, {24, 0x1}, {25, 0x1} } } },
+	[M_VTT_0_MPP_BUS]  = {"M_VTT_CTRL", 1, 1, {{{43, 0x2} } } },
+	[SDIO_0_MPP_BUS]  = {"SDIO 0", 10, 1, {{{48, 0x5}, {49, 0x5}, {50, 0x5}, {52, 0x5}, {53, 0x5}, {54, 0x5}, {55, 0x5}, {57, 0x5}, {58, 0x5}, {59, 0x5} } } },
 };
 
 struct mvebu_soc_info a38x_soc_info[] = {
diff --git a/arch/arm/include/asm/arch-mvebu/mpp.h b/arch/arm/include/asm/arch-mvebu/mpp.h
index e492cfd..b154b3f 100644
--- a/arch/arm/include/asm/arch-mvebu/mpp.h
+++ b/arch/arm/include/asm/arch-mvebu/mpp.h
@@ -24,7 +24,7 @@
 
 #define		MAX_MPP_NAME		32
 #define		MAX_BUS_NAME		32
-#define		MAX_PINS_PER_BUS	8
+#define		MAX_PINS_PER_BUS	14
 #define		MAX_BUS_OPTS		2
 
 char **mpp_get_desc_table(void);
@@ -36,6 +36,11 @@ enum mpp_bus_id {
 	SPI_1_MPP_BUS,
 	NAND_0_MPP_BUS,
 	RGMII_0_MPP_BUS,
+	I2C_0_MPP_BUS,
+	GE_0_MPP_BUS,
+	GE_1_MPP_BUS,
+	M_VTT_0_MPP_BUS,
+	SDIO_0_MPP_BUS,
 	MAX_MPP_BUS
 };
 
diff --git a/board/mvebu/armada38x/db-88f6820-info.c b/board/mvebu/armada38x/db-88f6820-info.c
index 1932a6f..cbe18d6 100644
--- a/board/mvebu/armada38x/db-88f6820-info.c
+++ b/board/mvebu/armada38x/db-88f6820-info.c
@@ -32,8 +32,8 @@ struct mvebu_board_config db_a38x_config = {
 struct mvebu_board_info a38x_db_info = {
 	.name = "DB-88F6820-BP",
 	.id = ARMADA_38X_DB_ID,
-	.mpp_regs = {0x11111111, 0x11111111, 0x55266011, 0x25252051,
-		     0x25255555, 0x40042565, 0x55550555, 0x00005550},
+	.mpp_regs = {0x11111111, 0x11111111, 0x11266011, 0x22222011,
+		     0x22200002, 0x40042022, 0x55550555, 0x00005550},
 	.configurable = 1,
 	.config_data = &db_a38x_config,
 };
diff --git a/board/mvebu/armada38x/rd-88f6820-info.c b/board/mvebu/armada38x/rd-88f6820-info.c
index 9df1f94..e610b54 100644
--- a/board/mvebu/armada38x/rd-88f6820-info.c
+++ b/board/mvebu/armada38x/rd-88f6820-info.c
@@ -40,8 +40,8 @@ struct mvebu_board_info a38x_rd_info = {
 	.id = ARMADA_38X_RD_ID,
 	.unit_mask = a38x_rd_unit_disable,
 	.unit_update_mode = UNIT_INFO_DISABLE,
-	.mpp_regs = {0x11111111, 0x11111111, 0x55266011, 0x25252051,
-		     0x25255555, 0x40042565, 0x55550555, 0x00005550},
+        .mpp_regs = {0x11111111, 0x11111111, 0x11266011, 0x22222011,
+		     0x22200002, 0x40042022, 0x55550555, 0x00005550},
 	.configurable = 1,
 	.config_data = &rd_a38x_config,
 };
diff --git a/board/mvebu/common/init.c b/board/mvebu/common/init.c
index 62b73a6..b19906f 100644
--- a/board/mvebu/common/init.c
+++ b/board/mvebu/common/init.c
@@ -144,7 +144,7 @@ int mvebu_board_init(void)
 
 #ifndef CONFIG_PALLADIUM
 	/* mpp_set */
-	mpp_enable_bus(UART_1_MPP_BUS, 0);
+	mpp_enable_bus(UART_0_MPP_BUS, 0);
 	mpp_set_and_update(brd->mpp_regs);
 #endif
 
-- 
1.9.1

